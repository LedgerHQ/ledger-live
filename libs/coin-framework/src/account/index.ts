export * from "./accountId";
export * from "./accountName";
export * from "./balanceHistoryCache";
export * from "./groupOperations";
export * from "./helpers";
export * from "./ordering";
export * from "./pending";
export * from "./serialization";
export * from "./support";
