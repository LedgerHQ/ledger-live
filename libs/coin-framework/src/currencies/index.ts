import { encodeURIScheme, decodeURIScheme } from "./CurrencyURIScheme";
import { sanitizeValueString } from "./sanitizeValueString";
import {
  listFiatCurrencies,
  findFiatCurrencyByTicker,
  getFiatCurrencyByTicker,
  hasFiatCurrencyTicker,
  listCryptoCurrencies,
  getCryptoCurrencyById,
  hasCryptoCurrencyId,
  findCryptoCurrency,
  findCryptoCurrencyById,
  findCryptoCurrencyByScheme,
  findCryptoCurrencyByKeyword,
  findCryptoCurrencyByTicker,
  listTokens,
  listTokensForCryptoCurrency,
  listTokenTypesForCryptoCurrency,
  findTokenByTicker,
  findTokenById,
  findTokenByAddress,
  hasTokenId,
  getAbandonSeedAddress,
  getTokenById,
  addTokens,
} from "@ledgerhq/cryptoassets";
export * from "./support";
import { parseCurrencyUnit } from "./parseCurrencyUnit";
import { chopCurrencyUnitDecimals } from "./chopCurrencyUnitDecimals";
import {
  formatCurrencyUnit,
  formatCurrencyUnitFragment,
} from "./formatCurrencyUnit";
import { formatShort } from "./formatShort";
import { valueFromUnit } from "./valueFromUnit";
import type { Currency } from "@ledgerhq/types-cryptoassets";

const findCurrencyByTicker = (ticker: string): Currency | null | undefined =>
  findCryptoCurrencyByTicker(ticker) ||
  findFiatCurrencyByTicker(ticker) ||
  findTokenByTicker(ticker);

export {
  listFiatCurrencies,
  listCryptoCurrencies,
  getFiatCurrencyByTicker,
  findCurrencyByTicker,
  findCryptoCurrency,
  findCryptoCurrencyById,
  findCryptoCurrencyByTicker,
  findCryptoCurrencyByScheme,
  findCryptoCurrencyByKeyword,
  findFiatCurrencyByTicker,
  hasFiatCurrencyTicker,
  listTokensForCryptoCurrency,
  listTokenTypesForCryptoCurrency,
  findTokenByAddress,
  findTokenByTicker,
  findTokenById,
  hasTokenId,
  getTokenById,
  getAbandonSeedAddress,
  parseCurrencyUnit,
  chopCurrencyUnitDecimals,
  formatCurrencyUnit,
  formatCurrencyUnitFragment,
  formatShort,
  getCryptoCurrencyById,
  hasCryptoCurrencyId,
  encodeURIScheme,
  decodeURIScheme,
  valueFromUnit,
  sanitizeValueString,
  listTokens,
  addTokens,
};
