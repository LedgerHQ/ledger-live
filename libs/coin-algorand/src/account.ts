import { getAccountUnit } from "@ledgerhq/coin-framework/account/index";
import { formatCurrencyUnit } from "@ledgerhq/coin-framework/currencies/index";
import type { Unit } from "@ledgerhq/types-cryptoassets";
import { BigNumber } from "bignumber.js";
import invariant from "invariant";
import type {
  AlgorandAccount,
  AlgorandOperation,
  AlgorandOperationExtra,
  AlgorandOperationExtraRaw,
} from "./types";

function formatOperationSpecifics(op: AlgorandOperation, unit: Unit | null | undefined): string {
  const { rewards } = op.extra;
  return rewards
    ? ` REWARDS : ${
        unit
          ? formatCurrencyUnit(unit, rewards, {
              showCode: true,
              disableRounding: true,
            }).padEnd(16)
          : rewards.toString()
      }`
    : "";
}

function formatAccountSpecifics(account: AlgorandAccount): string {
  const { algorandResources } = account;
  invariant(algorandResources, "algorand account expected");
  const unit = getAccountUnit(account);
  const formatConfig = {
    disableRounding: true,
    alwaysShowSign: false,
    showCode: true,
  };
  let str = " ";
  str += formatCurrencyUnit(unit, account.spendableBalance, formatConfig) + " spendable. ";

  if (algorandResources.rewards.gt(0)) {
    str += formatCurrencyUnit(unit, algorandResources.rewards, formatConfig) + " rewards. ";
  }

  return str;
}

export function fromOperationExtraRaw(extraRaw: AlgorandOperationExtraRaw) {
  const extra: AlgorandOperationExtra = {};
  if (extraRaw.rewards) {
    extra.rewards = new BigNumber(extraRaw.rewards);
  }

  if (extraRaw.memo) {
    extra.memo = extraRaw.memo;
  }

  if (extraRaw.assetId) {
    extra.assetId = extraRaw.assetId;
  }

  return extra;
}

export function toOperationExtraRaw(extra: AlgorandOperationExtra) {
  const extraRaw: AlgorandOperationExtraRaw = {};
  if (extra.rewards) {
    extraRaw.rewards = extra.rewards.toString();
  }

  if (extra.memo) {
    extraRaw.memo = extra.memo;
  }

  if (extra.assetId) {
    extraRaw.assetId = extra.assetId;
  }

  return extraRaw;
}

export default {
  formatAccountSpecifics,
  formatOperationSpecifics,
  fromOperationExtraRaw,
  toOperationExtraRaw,
};
