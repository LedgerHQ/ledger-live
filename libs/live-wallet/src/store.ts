/**
 * This exports all the logic related to the Wallet store.
 * The Wallet store is a store that contains the user data related to accounts.
 * It essentially is the whole user's wallet.
 */
import { Account, AccountLike, AccountRaw, AccountUserData } from "@ledgerhq/types-live";
import { getDefaultAccountName, getDefaultAccountNameForCurrencyIndex } from "./accountName";
import { AddAccountsAction } from "./addAccounts";
import { getCryptoCurrencyById } from "@ledgerhq/cryptoassets/currencies";
import { DistantState } from "./walletsync";
import { NonImportedAccountInfo } from "./walletsync/modules/accounts";

export type WSState = {
  data: DistantState | null;
  version: number;
};

export type WalletState = {
  // user's customized name for each account id
  accountNames: Map<string, string>;

  // a set of all the account ids that are starred (NB: token accounts can also be starred)
  starredAccountIds: Set<string>;

  nonImportedAccountInfos: NonImportedAccountInfo[];

  // local copy of the wallet sync data last synchronized with the backend of wallet sync, in order to be able to diff what we need to do when we apply an incremental update
  walletSyncState: WSState;
};

export type ExportedWalletState = {
  walletSyncState: WSState;
  nonImportedAccountInfos: NonImportedAccountInfo[];
};

export const initialState: WalletState = {
  accountNames: new Map(),
  starredAccountIds: new Set(),
  nonImportedAccountInfos: [],
  walletSyncState: { data: null, version: 0 },
};

export enum WalletHandlerType {
  INIT_ACCOUNTS = "INIT_ACCOUNTS",
  SET_ACCOUNT_NAME = "SET_ACCOUNT_NAME",
  SET_ACCOUNT_STARRED = "SET_ACCOUNT_STARRED",
  BULK_SET_ACCOUNT_NAMES = "BULK_SET_ACCOUNT_NAMES",
  ADD_ACCOUNTS = "ADD_ACCOUNTS",
  WALLET_SYNC_UPDATE = "WALLET_SYNC_UPDATE",
  IMPORT_WALLET_SYNC = "IMPORT_WALLET_SYNC",
  SET_NON_IMPORTED_ACCOUNTS = "SET_NON_IMPORTED_ACCOUNTS",
}

export type HandlersPayloads = {
  INIT_ACCOUNTS: { accounts: Account[]; accountsUserData: AccountUserData[] };
  SET_ACCOUNT_NAME: { accountId: string; name: string };
  BULK_SET_ACCOUNT_NAMES: { accountNames: Map<string, string> };
  SET_ACCOUNT_STARRED: { accountId: string; starred: boolean };
  ADD_ACCOUNTS: AddAccountsAction["payload"];
  WALLET_SYNC_UPDATE: {
    data: DistantState | null;
    version: number;
  };
  IMPORT_WALLET_SYNC: Partial<ExportedWalletState>;
  SET_NON_IMPORTED_ACCOUNTS: NonImportedAccountInfo[];
};

type Handlers<State, Types, PreciseKey = true> = {
  [Key in keyof Types]: (
    state: State,
    body: { payload: Types[PreciseKey extends true ? Key : keyof Types] },
  ) => State;
};

export type WalletHandlers<PreciseKey = true> = Handlers<WalletState, HandlersPayloads, PreciseKey>;

export const handlers: WalletHandlers = {
  INIT_ACCOUNTS: (state, { payload: { accountsUserData } }): WalletState => {
    const accountNames = new Map();
    const starredAccountIds = new Set<string>();
    accountsUserData.forEach(accountUserData => {
      accountNames.set(accountUserData.id, accountUserData.name);
      for (const starredId of accountUserData.starredIds) {
        starredAccountIds.add(starredId);
      }
    });
    return {
      ...state,
      accountNames,
      starredAccountIds,
    };
  },
  SET_ACCOUNT_NAME: (state, { payload: { accountId, name } }) => {
    const accountNames = new Map(state.accountNames);
    if (!name) {
      accountNames.delete(accountId);
    } else {
      accountNames.set(accountId, name);
    }
    return { ...state, accountNames };
  },
  BULK_SET_ACCOUNT_NAMES: (state, { payload: { accountNames } }) => {
    // merge the new account names with the existing ones
    const newAccountNames = new Map(state.accountNames);
    for (const [accountId, name] of accountNames) {
      newAccountNames.set(accountId, name);
    }
    return { ...state, accountNames: newAccountNames };
  },
  SET_ACCOUNT_STARRED: (state, { payload: { accountId, starred: value } }) => {
    const starredAccountIds = new Set(state.starredAccountIds);
    if (value) {
      starredAccountIds.add(accountId);
    } else {
      starredAccountIds.delete(accountId);
    }
    return { ...state, starredAccountIds };
  },
  ADD_ACCOUNTS: (state, { payload: { allAccounts, editedNames } }) => {
    const accountNames = new Map(state.accountNames);
    for (const account of allAccounts) {
      const name =
        editedNames.get(account.id) ||
        accountNames.get(account.id) ||
        getDefaultAccountName(account);
      accountNames.set(account.id, name);
    }
    return { ...state, accountNames };
  },
  WALLET_SYNC_UPDATE: (state, { payload }) => {
    return { ...state, walletSyncState: payload };
  },
  IMPORT_WALLET_SYNC: (state, { payload }) => {
    return {
      ...state,
      ...payload,
    };
  },
  SET_NON_IMPORTED_ACCOUNTS: (state, { payload }) => {
    return { ...state, nonImportedAccountInfos: payload };
  },
};

// actions

export const setAccountName = (accountId: string, name: string) => ({
  type: "SET_ACCOUNT_NAME",
  payload: { accountId, name },
});

export const setAccountNames = (accountNames: Map<string, string>) => ({
  type: "BULK_SET_ACCOUNT_NAMES",
  payload: { accountNames },
});

export const setAccountStarred = (accountId: string, starred: boolean) => ({
  type: "SET_ACCOUNT_STARRED",
  payload: { accountId, starred },
});

export const initAccounts = (accounts: Account[], accountsUserData: AccountUserData[]) => ({
  type: "INIT_ACCOUNTS",
  payload: { accounts, accountsUserData },
});

/**
 * action to import back the wallet state. opposite of exportWalletState
 */
export const importWalletState = (payload: Partial<ExportedWalletState>) => ({
  type: "IMPORT_WALLET_SYNC",
  payload,
});

export const walletSyncUpdate = (data: DistantState | null, version: number) => ({
  type: "WALLET_SYNC_UPDATE",
  payload: { data, version },
});

export const setNonImportedAccounts = (payload: NonImportedAccountInfo[]) => ({
  type: "SET_NON_IMPORTED_ACCOUNTS",
  payload,
});

// Local Selectors

export const accountNameSelector = (
  state: WalletState,
  { accountId }: { accountId: string },
): string | undefined => state.accountNames.get(accountId);

export const accountNameWithDefaultSelector = (state: WalletState, account: AccountLike): string =>
  state.accountNames.get(account.id) || getDefaultAccountName(account);

export const isStarredAccountSelector = (
  state: WalletState,
  { accountId }: { accountId: string },
): boolean => state.starredAccountIds.has(accountId);

/**
 * Recreate an AccountUserData from the store.
 * it is used to transport all data related to a main account.
 * but the data isn't used internally for performance reason.
 */
export const accountUserDataExportSelector = (
  state: WalletState,
  { account }: { account: Account },
): AccountUserData => {
  const id = account.id;
  const name = state.accountNames.get(id) || getDefaultAccountName(account);
  const starredIds = [];
  if (state.starredAccountIds.has(id)) {
    starredIds.push(id);
  }
  for (const t of account.subAccounts || []) {
    if (state.starredAccountIds.has(t.id)) {
      starredIds.push(t.id);
    }
  }
  return { id, name, starredIds };
};

export const accountRawToAccountUserData = (raw: AccountRaw): AccountUserData => {
  const { id } = raw;
  const name =
    raw.name ||
    getDefaultAccountNameForCurrencyIndex({
      currency: getCryptoCurrencyById(raw.currencyId),
      index: raw.index,
    });
  const starredIds = [];
  if (raw.starred) {
    starredIds.push(raw.id);
  }
  for (const t of raw.subAccounts || []) {
    if (t.starred) {
      starredIds.push(t.id);
    }
  }
  return { id, name, starredIds };
};

/**
 * call this selector to save the store state
 */
export const exportWalletState = (state: WalletState): ExportedWalletState => ({
  walletSyncState: state.walletSyncState,
  nonImportedAccountInfos: state.nonImportedAccountInfos,
});

export const walletStateExportShouldDiffer = (a: WalletState, b: WalletState): boolean => {
  return (
    a.walletSyncState !== b.walletSyncState ||
    a.nonImportedAccountInfos !== b.nonImportedAccountInfos
  );
};

export const walletSyncStateSelector = (state: WalletState): WSState => state.walletSyncState;
