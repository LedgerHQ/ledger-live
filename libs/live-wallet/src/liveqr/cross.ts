// cross helps dealing with cross-project feature like export/import & cross project conversions
import { BigNumber } from "bignumber.js";
// @ts-expect-error no types for compressjs
import compressjs from "@ledgerhq/compressjs";
import type { DeviceModelId } from "@ledgerhq/devices";
import {
  runDerivationScheme,
  getDerivationScheme,
  asDerivationMode,
} from "@ledgerhq/coin-framework/derivation";
import { decodeAccountId, emptyHistoryCache } from "@ledgerhq/coin-framework/account/index";
import { getCryptoCurrencyById } from "@ledgerhq/cryptoassets";
import type { Account, AccountUserData, CryptoCurrencyIds } from "@ledgerhq/types-live";
import { WalletState, accountUserDataExportSelector } from "../store";

export type AccountData = {
  id: string;
  currencyId: string;
  freshAddress?: string;
  seedIdentifier: string;
  derivationMode: string;
  // we are unsafe at this stage, validation is done later
  name: string;
  index: number;
  balance: string;
};

export type CryptoSettings = {
  confirmationsNb?: number;
};

export type Settings = {
  counterValue?: string;
  currenciesSettings: Record<CryptoCurrencyIds, CryptoSettings>;
  pairExchanges: Record<string, string | null | undefined>;
  blacklistedTokenIds?: string[];
  hideEmptyTokenAccounts?: boolean;
};

export type DataIn = {
  // wallet state
  walletState: WalletState;
  // accounts to export (filter them to only be the visible ones)
  accounts: Account[];
  // settings
  settings: Settings;
  // the name of the exporter. e.g. "desktop" for the desktop app
  exporterName: string;
  // the version of the exporter. e.g. the desktop app version
  exporterVersion: string;
  modelId?: DeviceModelId;
  modelIdList?: DeviceModelId[];
};

type Meta = {
  exporterName: string;
  exporterVersion: string;
  modelId?: DeviceModelId;
  modelIdList?: DeviceModelId[];
};

export type Result = {
  accounts: AccountData[];
  settings: Settings;
  meta: Meta;
};

export function encode({
  walletState,
  accounts,
  settings,
  exporterName,
  exporterVersion,
  modelId,
  modelIdList,
}: DataIn): string {
  return Buffer.from(
    compressjs.Bzip2.compressFile(
      Buffer.from(
        JSON.stringify({
          meta: {
            exporterName,
            exporterVersion,
            modelId,
            modelIdList,
          },
          accounts: accounts.map(a =>
            accountToAccountData(a, accountUserDataExportSelector(walletState, { account: a })),
          ),
          settings,
        }),
      ),
    ),
  ).toString("binary");
}

const asResultMeta = (unsafe: Record<string, unknown>): Meta => {
  if (typeof unsafe !== "object" || !unsafe) {
    throw new Error("invalid meta data");
  }

  const { exporterName, exporterVersion, modelId, modelIdList } = unsafe;

  if (typeof exporterName !== "string") {
    throw new Error("invalid meta.exporterName");
  }

  if (typeof exporterVersion !== "string") {
    throw new Error("invalid meta.exporterVersion");
  }

  let resultModelId: DeviceModelId | undefined = undefined;
  if (modelId) {
    if (typeof modelId === "string") {
      resultModelId = modelId as DeviceModelId;
    } else {
      throw new Error("invalid meta.modelId");
    }
  }

  let resultModelIdList: DeviceModelId[] | undefined = undefined;

  if (modelIdList) {
    if (
      typeof modelIdList === "object" &&
      Array.isArray(modelIdList) &&
      modelIdList.every(id => typeof id === "string")
    ) {
      resultModelIdList = modelIdList as DeviceModelId[];
    } else {
      throw new Error("invalid meta.modelIdList");
    }
  }

  return {
    exporterName,
    exporterVersion,
    modelId: resultModelId,
    modelIdList: resultModelIdList,
  };
};

const asResultAccount = (unsafe: Record<string, unknown>): AccountData => {
  if (typeof unsafe !== "object" || !unsafe) {
    throw new Error("invalid account data");
  }

  const { id, currencyId, freshAddress, seedIdentifier, derivationMode, name, index, balance } =
    unsafe;

  if (typeof id !== "string") {
    throw new Error("invalid account.id");
  }

  if (typeof currencyId !== "string") {
    throw new Error("invalid account.currencyId");
  }

  if (typeof seedIdentifier !== "string") {
    throw new Error("invalid account.seedIdentifier");
  }

  if (typeof derivationMode !== "string") {
    throw new Error("invalid account.derivationMode");
  }

  if (typeof name !== "string") {
    throw new Error("invalid account.name");
  }

  if (typeof index !== "number") {
    throw new Error("invalid account.index");
  }

  if (typeof balance !== "string") {
    throw new Error("invalid account.balance");
  }

  const o: AccountData = {
    id,
    currencyId,
    seedIdentifier,
    derivationMode,
    name,
    index,
    balance,
  };

  if (typeof freshAddress === "string" && freshAddress) {
    o.freshAddress = freshAddress;
  }

  return o;
};

const asResultAccounts = (unsafe: unknown): AccountData[] => {
  if (typeof unsafe !== "object" || !unsafe || !Array.isArray(unsafe)) {
    throw new Error("invalid accounts data");
  }

  return unsafe.map(asResultAccount);
};

const asCryptoSettings = (unsafe: Record<string, unknown>): CryptoSettings => {
  if (typeof unsafe !== "object" || !unsafe) {
    throw new Error("invalid currency settings data");
  }

  const { confirmationsNb } = unsafe;

  if (typeof confirmationsNb === "number") {
    return {
      confirmationsNb,
    };
  }

  return {};
};

const asResultSettings = (unsafe: Record<string, any>): Settings => {
  if (typeof unsafe !== "object" || !unsafe) {
    throw new Error("invalid settings data");
  }

  const {
    counterValue,
    currenciesSettings,
    pairExchanges,
    blacklistedTokenIds,
    hideEmptyTokenAccounts,
  } = unsafe;
  const currenciesSettingsSafe: Record<CryptoCurrencyIds, CryptoSettings> = {};

  if (currenciesSettings && typeof currenciesSettings === "object") {
    for (const k in currenciesSettings) {
      currenciesSettingsSafe[k] = asCryptoSettings(currenciesSettings[k]);
    }
  }

  const pairExchangesSafe: Record<string, string> = {};

  if (pairExchanges && typeof pairExchanges === "object") {
    for (const k in pairExchanges) {
      const v = pairExchanges[k];

      if (v && typeof v === "string") {
        pairExchangesSafe[k] = v;
      }
    }
  }

  const res: Settings = {
    currenciesSettings: currenciesSettingsSafe,
    pairExchanges: pairExchangesSafe,
  };

  if (counterValue && typeof counterValue === "string") {
    res.counterValue = counterValue;
  }

  if (hideEmptyTokenAccounts && typeof hideEmptyTokenAccounts === "boolean") {
    res.hideEmptyTokenAccounts = hideEmptyTokenAccounts;
  }

  const blacklistedTokenIdsSafe: string[] = [];

  if (blacklistedTokenIds && Array.isArray(blacklistedTokenIds)) {
    for (const b of blacklistedTokenIds) {
      if (typeof b === "string") {
        blacklistedTokenIdsSafe.push(b);
      }
    }

    res.blacklistedTokenIds = blacklistedTokenIdsSafe;
  }

  return res;
};

export function decode(bytes: string): Result {
  const unsafe: Record<string, any> = JSON.parse(
    Buffer.from(compressjs.Bzip2.decompressFile(Buffer.from(bytes, "binary"))).toString(),
  );

  if (typeof unsafe !== "object" || !unsafe) {
    throw new Error("invalid data");
  }

  return {
    meta: asResultMeta(unsafe.meta),
    accounts: asResultAccounts(unsafe.accounts),
    settings: asResultSettings(unsafe.settings),
  };
}
export function accountToAccountData(
  { id, seedIdentifier, derivationMode, freshAddress, currency, index, balance }: Account,
  { name }: AccountUserData,
): AccountData {
  const res: AccountData = {
    id,
    name,
    seedIdentifier,
    derivationMode,
    freshAddress,
    currencyId: currency.id,
    index,
    balance: balance.toString(),
  };

  return res;
}
// reverse the account data to an account.
// this restore the essential data of an account and the result of the fields
// are assumed to be restored during first sync
export const accountDataToAccount = ({
  id,
  currencyId,
  freshAddress: inputFreshAddress,
  name,
  index,
  balance,
  derivationMode: derivationModeStr,
  seedIdentifier,
}: AccountData): [Account, AccountUserData] => {
  const { xpubOrAddress } = decodeAccountId(id); // TODO rename in AccountId xpubOrAddress

  const derivationMode = asDerivationMode(derivationModeStr);
  const currency = getCryptoCurrencyById(currencyId);
  let xpub = "";
  let freshAddress = inputFreshAddress || "";
  let freshAddressPath = "";

  if (
    // FIXME Dirty hack, since we have no way here to know if "xpubOrAddress" is one or the other.
    // Proposed fix: https://ledgerhq.atlassian.net/browse/LL-7437
    currency.family === "bitcoin" ||
    currency.family === "cardano"
  ) {
    // In bitcoin implementation, xpubOrAddress field always go in the xpub
    xpub = xpubOrAddress;
  } else {
    if (currency.family === "tezos" || currency.family === "stacks") {
      xpub = xpubOrAddress;
    } else if (!freshAddress) {
      // otherwise, it's the freshAddress
      freshAddress = xpubOrAddress;
    }

    freshAddressPath = runDerivationScheme(
      getDerivationScheme({
        currency,
        derivationMode,
      }),
      currency,
      {
        account: index,
      },
    );
  }

  const balanceBN = new BigNumber(balance);
  const account: Account = {
    type: "Account",
    id,
    derivationMode,
    seedIdentifier,
    xpub,
    used: false,
    currency,
    index,
    freshAddress,
    freshAddressPath,
    swapHistory: [],
    blockHeight: 0,
    balance: balanceBN,
    spendableBalance: balanceBN,
    operationsCount: 0,
    operations: [],
    pendingOperations: [],
    lastSyncDate: new Date(0),
    creationDate: new Date(),
    balanceHistoryCache: emptyHistoryCache,
  };

  const accountUserData: AccountUserData = {
    id,
    name,
    starredIds: [],
  };

  return [account, accountUserData];
};
