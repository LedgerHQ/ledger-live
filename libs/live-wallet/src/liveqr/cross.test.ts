import { genAccount } from "@ledgerhq/coin-framework/mocks/account";
import { getDerivationModesForCurrency } from "@ledgerhq/coin-framework/derivation";
import { listCryptoCurrencies } from "@ledgerhq/cryptoassets/index";
import { setSupportedCurrencies } from "@ledgerhq/coin-framework/currencies/index";
import { accountDataToAccount, accountToAccountData } from "./cross";
import { accountUserDataExportSelector, initialState } from "../store";

setSupportedCurrencies(["ethereum", "ethereum_classic"]);

test("accountDataToAccount / accountToAccountData", () => {
  listCryptoCurrencies().forEach(currency => {
    getDerivationModesForCurrency(currency).forEach(derivationMode => {
      const account = genAccount(`${currency.id}_${derivationMode}`, { currency });
      const walletState = initialState;
      const data = accountToAccountData(
        account,
        accountUserDataExportSelector(walletState, { account }),
      );
      expect(accountToAccountData(...accountDataToAccount(data))).toMatchObject(data);
    });
  });
});
