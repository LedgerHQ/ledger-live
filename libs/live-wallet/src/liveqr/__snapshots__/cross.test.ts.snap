// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`encode/decode 1`] = `
{
  "accounts": [
    {
      "balance": "349005855825536773784",
      "currencyId": "ethereum_classic",
      "derivationMode": "",
      "freshAddress": "0xF882CE5EF3E667AAFF592AE27793E0CF40356D1E",
      "id": "mock:1:ethereum_classic:export_0:",
      "index": 1,
      "name": "uuwygRL0AmMrs9riAlv1",
      "seedIdentifier": "mock",
    },
    {
      "balance": "45503129837786898140",
      "currencyId": "ethereum_classic",
      "derivationMode": "",
      "freshAddress": "0x22FF1149E5C7585F99647784C2D2A8D6415979F3",
      "id": "mock:1:ethereum_classic:export_1:",
      "index": 1,
      "name": "4GEb2iRQXlUt6X1jviKzSz4umwWj",
      "seedIdentifier": "mock",
    },
    {
      "balance": "288566332368580450524",
      "currencyId": "ethereum_classic",
      "derivationMode": "",
      "freshAddress": "0x96B35855C525CA59FEEC9F1F88B6A59645C0034D",
      "id": "mock:0:ethereum_classic:export_2:",
      "index": 1,
      "name": "tzPbPqv6XS7BabfsuJff",
      "seedIdentifier": "mock",
    },
  ],
  "meta": {
    "exporterName": "test",
    "exporterVersion": "0.0.0",
    "modelId": undefined,
    "modelIdList": undefined,
  },
  "settings": {
    "blacklistedTokenIds": [
      "tokenid1",
      "tokenid2",
    ],
    "counterValue": "USD",
    "currenciesSettings": {
      "bitcoin": {
        "confirmationsNb": 3,
      },
    },
    "pairExchanges": {
      "BTC_USD": "KRAKEN",
    },
  },
}
`;
